core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala

```diff
@@ -0,0 +1,442 @@
+          // Note that this line won't be invoked (and thus we won't incur the update() cost)
+          // when endingVersion is present.
+          deltaLog.update().version
+        },
+        sqlContext.sparkSession)
+
+      df.select(requiredColumns.map(SchemaUtils.fieldNameToColumn): _*).rdd
+    }
+  }
+
+  /**
+   * For a sequence of changes(AddFile, RemoveFile, AddCDCFile) create a DataFrame that represents
+   * that captured change data between start and end inclusive.
+   *
+   * @param deltaLog - DeltaLog for the table for which we are creating a cdc dataFrame
+   * @param start - startingVersion of the changes
+   * @param end - endingVersion of the changes
+   * @param changes - changes is an iterator of all FileActions for a particular commit version.
+   * @param isStreaming - indicates whether the DataFrame returned is a streaming DataFrame
+   * @param spark - SparkSession
+   * @return CDCInfo which contains the DataFrame of the changes as well as the statistics
+   *         related to the changes
+   */
+  def changesToDF(
+      deltaLog: DeltaLog,
+      start: Long,
+      end: Long,
+      changes: Iterator[(Long, Seq[Action])],
+      spark: SparkSession,
+      isStreaming: Boolean = false): CDCVersionDiffInfo = {
+
+    if (end < start) {
+      throw DeltaErrors.endBeforeStartVersionInCDC(start, end)
+    }
+
+    val snapshot = deltaLog.snapshot
+
+    // Correct timestamp values are only available through DeltaHistoryManager.getCommits(). Commit
+    // info timestamps are wrong, and file modification times are wrong because they need to be
+    // monotonized first. This just performs a list (we don't read the contents of the files in
+    // getCommits()) so it's not a big deal.
+    val timestampsByVersion: Map[Long, Timestamp] = {
+      val monotonizationStart =
+        Seq(start - DeltaHistoryManager.POTENTIALLY_UNMONOTONIZED_TIMESTAMPS, 0).max
+      val commits = DeltaHistoryManager.getCommits(
+        deltaLog.store,
+        deltaLog.logPath,
+        monotonizationStart,
+        Some(end + 1),
+        deltaLog.newDeltaHadoopConf())
+
+      // Note that the timestamps come from filesystem modification timestamps, so they're
+      // milliseconds since epoch and we don't need to deal with timezones.
+      commits.map(f => (f.version -> new Timestamp(f.timestamp))).toMap
+    }
+
+    val changeFiles = ListBuffer[CDCDataSpec[AddCDCFile]]()
+    val addFiles = ListBuffer[CDCDataSpec[AddFile]]()
+    val removeFiles = ListBuffer[CDCDataSpec[RemoveFile]]()
+    if (!isCDCEnabledOnTable(deltaLog.getSnapshotAt(start).metadata)) {
+      throw DeltaErrors.changeDataNotRecordedException(start, start, end)
+    }
+
+    var totalBytes = 0L
+    var totalFiles = 0L
+
+    changes.foreach {
+      case (v, actions) =>
+        // Check whether CDC was newly disabled in this version. (We should have already checked
```

Comment:
nit: Why is this line in a bracket

Reply:
The first sentence is the important info. The sentence in the bracket is extra info. Just my style of writing.